v3.13.7 plugin crashed

Hi,

Code was running fine yesterday. Today ran terraform plan on same code base on got following output:

Stack trace from the terraform-provider-okta_v3.13.7 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x18ad682]

goroutine 107 [running]:
github.com/okta/terraform-provider-okta/okta.resourcePolicyRuleIdpDiscoveryRead(0x1bf0528, 0xc00053c7e0, 0xc00030c200, 0x199a8e0, 0xc000270ea0, 0xc000657ad0, 0xc0001bf908, 0x100df38)
	github.com/okta/terraform-provider-okta/okta/resource_okta_policy_rule_idp_discovery.go:110 +0x522
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0xc0004577a0, 0x1bf04b8, 0xc0003a21c0, 0xc00030c200, 0x199a8e0, 0xc000270ea0, 0x0, 0x0, 0x0)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.0/helper/schema/resource.go:347 +0x17f
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc0004577a0, 0x1bf04b8, 0xc0003a21c0, 0xc0003761c0, 0x199a8e0, 0xc000270ea0, 0xc000626390, 0x0, 0x0, 0x0)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.0/helper/schema/resource.go:624 +0x1cb
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadResource(0xc000490a20, 0x1bf04b8, 0xc0003a21c0, 0xc0003a2200, 0xc0003a21c0, 0x100b665, 0x1a41740)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.0/helper/schema/grpc_provider.go:575 +0x43b
github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).ReadResource(0xc000149d40, 0x1bf0560, 0xc0003a21c0, 0xc00053c300, 0xc000149d40, 0xc000650a20, 0xc00053eba0)
	github.com/hashicorp/terraform-plugin-go@v0.3.0/tfprotov5/server/server.go:298 +0x105
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadResource_Handler(0x1a77280, 0xc000149d40, 0x1bf0560, 0xc000650a20, 0xc00053c2a0, 0x0, 0x1bf0560, 0xc000650a20, 0xc000818200, 0x1e2)
	github.com/hashicorp/terraform-plugin-go@v0.3.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:344 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000232c40, 0x1bf8c78, 0xc000103980, 0xc0000ee100, 0xc0003004b0, 0x20d4830, 0x0, 0x0, 0x0)
	google.golang.org/grpc@v1.32.0/server.go:1194 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc000232c40, 0x1bf8c78, 0xc000103980, 0xc0000ee100, 0x0)
	google.golang.org/grpc@v1.32.0/server.go:1517 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000302050, 0xc000232c40, 0x1bf8c78, 0xc000103980, 0xc0000ee100)
	google.golang.org/grpc@v1.32.0/server.go:859 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.32.0/server.go:857 +0x1fd

Error: The terraform-provider-okta_v3.13.7 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.


<!--- Please keep this note for the community --->

### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform Version

<!--- Please run `terraform -v` to show the Terraform core version and provider version(s). If you are not running the latest version of Terraform or the provider, please upgrade because your issue may have already been fixed. [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions). --->

### Affected Resource(s)

<!--- Please list the affected resources and data sources. --->

- okta_xxx

### Terraform Configuration Files

<!--- Information about code formatting: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code --->

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key: https://keybase.io/hashicorp
```

### Debug Output

<!---
Please provide a link to a GitHub Gist containing the complete debug output. Please do NOT paste the debug output in the issue; just paste a link to the Gist.

To obtain the debug output, see the [Terraform documentation on debugging](https://www.terraform.io/docs/internals/debugging.html).
--->

### Panic Output

<!--- If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`. --->

### Expected Behavior

<!--- What should have happened? --->

### Actual Behavior

<!--- What actually happened? --->

### Steps to Reproduce

<!--- Please list the steps required to reproduce the issue. --->

1. `terraform apply`

### Important Factoids

<!--- Are there anything atypical about your accounts that we should know? For example: Running in EC2 Classic? --->

### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Vendor documentation? For example:
--->

- #0000

--- a/examples/okta_user/all_attributes.tf
+++ b/examples/okta_user/all_attributes.tf
@@ -1,5 +1,4 @@
 resource "okta_user" "test" {
-  admin_roles        = ["ORG_ADMIN"]
   first_name         = "TestAcc"
   last_name          = "Smith"
   login              = "testAcc-replace_with_uuid@example.com"

--- a/examples/okta_user/custom_attributes.tf
+++ b/examples/okta_user/custom_attributes.tf
@@ -23,7 +23,6 @@ resource "okta_user_schema_property" "test_number" {
 }
 
 resource "okta_user" "test" {
-  admin_roles = ["APP_ADMIN", "USER_ADMIN"]
   first_name  = "TestAcc"
   last_name   = "Smith"
   login       = "testAcc-replace_with_uuid@example.com"

--- a/examples/okta_user/custom_attributes_array.tf
+++ b/examples/okta_user/custom_attributes_array.tf
@@ -23,7 +23,6 @@ resource "okta_user_schema_property" "test_number" {
 }
 
 resource "okta_user" "test" {
-  admin_roles = ["APP_ADMIN", "USER_ADMIN"]
   first_name  = "TestAcc"
   last_name   = "Smith"
   login       = "testAcc-replace_with_uuid@example.com"

--- a/examples/okta_user/password_hash.tf
+++ b/examples/okta_user/password_hash.tf
@@ -1,0 +1,13 @@
+resource "okta_user" "test" {
+  first_name = "TestAcc"
+  last_name  = "Smith"
+  login      = "testAcc-replace_with_uuid@example.com"
+  email      = "testAcc-replace_with_uuid@example.com"
+  status     = "STAGED"
+  password_hash {
+    algorithm  = "SHA-512"
+    salt       = "TXlTYWx0"
+    salt_order = "PREFIX"
+    value      = "QrozP8a+KfoHu6mPFysxLoO5LMQsd2Fw6IclZUf8xQjetJOCGS93vm68h+VaFX0LHSiF/GxQkykq1vofmx6NGA=="
+  }
+}

--- a/examples/okta_user/password_hash_updated.tf
+++ b/examples/okta_user/password_hash_updated.tf
@@ -1,0 +1,13 @@
+resource "okta_user" "test" {
+  first_name = "TestAcc"
+  last_name  = "Smith"
+  login      = "testAcc-replace_with_uuid@example.com"
+  email      = "testAcc-replace_with_uuid@example.com"
+  status     = "STAGED"
+  password_hash {
+    algorithm   = "BCRYPT"
+    work_factor = 10
+    salt        = "rwh3vH166HCH/NT9XV5FYu"
+    value       = "qaMqvAPULkbiQzkTCWo5XDcvzpk8Tna"
+  }
+}

--- a/examples/okta_user/remove_custom_attributes.tf
+++ b/examples/okta_user/remove_custom_attributes.tf
@@ -1,5 +1,4 @@
 resource "okta_user" "test" {
-  admin_roles = ["APP_ADMIN", "USER_ADMIN"]
   first_name  = "TestAcc"
   last_name   = "Smith"
   login       = "testAcc-replace_with_uuid@example.com"

--- a/examples/okta_user/staged.tf
+++ b/examples/okta_user/staged.tf
@@ -1,5 +1,4 @@
 resource "okta_user" "test" {
-  admin_roles = ["APP_ADMIN", "USER_ADMIN"]
   first_name  = "TestAcc"
   last_name   = "Smith"
   login       = "testAcc-replace_with_uuid@example.com"

--- a/okta/app.go
+++ b/okta/app.go
@@ -555,7 +555,7 @@ func setSamlSettings(d *schema.ResourceData, signOn *okta.SamlApplicationSetting
 				for i := range acsEndpointsObj {
 					acsEndpoints[i] = acsEndpointsObj[i].Url
 				}
-				_ = d.Set("acs_endpoints", convertStringSliceToSet(acsEndpoints))
+				_ = d.Set("acs_endpoints", convertStringSliceToSetNullable(acsEndpoints))
 			}
 		} else {
 			_ = d.Set("acs_endpoints", nil)

--- a/okta/data_source_okta_app_saml.go
+++ b/okta/data_source_okta_app_saml.go
@@ -358,7 +358,7 @@ func dataSourceAppSamlRead(ctx context.Context, d *schema.ResourceData, m interf
 			return diag.Errorf("failed to read SAML app: failed to set SAML app settings: %v", err)
 		}
 	}
-	_ = d.Set("features", convertStringSliceToSet(app.Features))
+	_ = d.Set("features", convertStringSliceToSetNullable(app.Features))
 	_ = d.Set("user_name_template", app.Credentials.UserNameTemplate.Template)
 	_ = d.Set("user_name_template_type", app.Credentials.UserNameTemplate.Type)
 	_ = d.Set("user_name_template_suffix", app.Credentials.UserNameTemplate.Suffix)

--- a/okta/policy_rule.go
+++ b/okta/policy_rule.go
@@ -200,12 +200,17 @@ func syncRuleFromUpstream(d *schema.ResourceData, rule *sdk.PolicyRule) error {
 	_ = d.Set("status", rule.Status)
 	_ = d.Set("priority", rule.Priority)
 	_ = d.Set("network_connection", rule.Conditions.Network.Connection)
+	m := map[string]interface{}{
+		"users_excluded": convertStringSliceToSetNullable(rule.Conditions.People.Users.Exclude),
+	}
+	if len(rule.Conditions.Network.Include) > 0 {
+		m["network_includes"] = convertStringSliceToInterfaceSlice(rule.Conditions.Network.Include)
+	}
+	if len(rule.Conditions.Network.Exclude) > 0 {
+		m["network_excludes"] = convertStringSliceToInterfaceSlice(rule.Conditions.Network.Exclude)
+	}
 	if rule.Conditions.Network.Connection != "ANYWHERE" {
-		return setNonPrimitives(d, map[string]interface{}{
-			"users_excluded":   convertStringSliceToSetNullable(rule.Conditions.People.Users.Exclude),
-			"network_includes": convertStringArrToInterface(rule.Conditions.Network.Include),
-			"network_excludes": convertStringArrToInterface(rule.Conditions.Network.Exclude),
-		})
+		return setNonPrimitives(d, m)
 	}
 	return setNonPrimitives(d, map[string]interface{}{
 		"users_excluded": convertStringSliceToSetNullable(rule.Conditions.People.Users.Exclude),

--- a/okta/resource_okta_app_group_assignments.go
+++ b/okta/resource_okta_app_group_assignments.go
@@ -18,14 +18,12 @@ func resourceAppGroupAssignments() *schema.Resource {
 		DeleteContext: resourceAppGroupAssignmentsDelete,
 		UpdateContext: resourceAppGroupAssignmentsUpdate,
 		Importer:      &schema.ResourceImporter{},
-
 		Schema: map[string]*schema.Schema{
 			"app_id": {
 				Type:     schema.TypeString,
 				Required: true,
 				ForceNew: true,
 			},
-
 			"group": {
 				Type:        schema.TypeSet,
 				Required:    true,

--- a/okta/resource_okta_domain.go
+++ b/okta/resource_okta_domain.go
@@ -109,7 +109,7 @@ func resourceDomainRead(ctx context.Context, d *schema.ResourceData, m interface
 			"expiration":  domain.DnsRecords[i].Expiration,
 			"fqdn":        domain.DnsRecords[i].Fqdn,
 			"record_type": domain.DnsRecords[i].RecordType,
-			"values":      convertStringArrToInterface(domain.DnsRecords[i].Values),
+			"values":      convertStringSliceToInterfaceSlice(domain.DnsRecords[i].Values),
 		}
 	}
 	err = setNonPrimitives(d, map[string]interface{}{"dns_records": arr})

--- a/okta/resource_okta_policy_rule_idp_discovery.go
+++ b/okta/resource_okta_policy_rule_idp_discovery.go
@@ -109,15 +109,20 @@ func resourcePolicyRuleIdpDiscoveryRead(ctx context.Context, d *schema.ResourceD
 	_ = d.Set("priority", rule.Priority)
 	_ = d.Set("user_identifier_attribute", rule.Conditions.UserIdentifier.Attribute)
 	_ = d.Set("user_identifier_type", rule.Conditions.UserIdentifier.Type)
-	_ = d.Set("network_connection", rule.Conditions.Network.Connection)
-	err = setNonPrimitives(d, map[string]interface{}{
-		"network_includes":         convertStringArrToInterface(rule.Conditions.Network.Include),
-		"network_excludes":         convertStringArrToInterface(rule.Conditions.Network.Exclude),
+	mm := map[string]interface{}{
 		"platform_include":         flattenPlatformInclude(rule.Conditions.Platform),
-		"user_identifier_patterns": flattenUserIDPatterns(rule.Conditions.UserIdentifier.Patterns),
 		"app_include":              flattenAppInclude(rule.Conditions.App),
 		"app_exclude":              flattenAppExclude(rule.Conditions.App),
-	})
+		"user_identifier_patterns": flattenUserIDPatterns(rule.Conditions.UserIdentifier.Patterns),
+	}
+	_ = d.Set("network_connection", rule.Conditions.Network.Connection)
+	if len(rule.Conditions.Network.Include) > 0 {
+		mm["network_includes"] = convertStringSliceToInterfaceSlice(rule.Conditions.Network.Include)
+	}
+	if len(rule.Conditions.Network.Exclude) > 0 {
+		mm["network_excludes"] = convertStringSliceToInterfaceSlice(rule.Conditions.Network.Exclude)
+	}
+	err = setNonPrimitives(d, mm)
 	if err != nil {
 		return diag.Errorf("failed to set IDP discovery policy rule properties: %v", err)
 	}

--- a/okta/resource_okta_user.go
+++ b/okta/resource_okta_user.go
@@ -61,12 +61,15 @@ func resourceUser() *schema.Resource {
 		Importer: &schema.ResourceImporter{
 			StateContext: func(ctx context.Context, d *schema.ResourceData, m interface{}) ([]*schema.ResourceData, error) {
 				// Supporting id and email based imports
-				client := getOktaClientFromMetadata(m)
-				user, _, err := client.User.GetUser(ctx, d.Id())
+				user, _, err := getOktaClientFromMetadata(m).User.GetUser(ctx, d.Id())
 				if err != nil {
 					return nil, err
 				}
 				d.SetId(user.Id)
+				err = setAdminRoles(ctx, d, m)
+				if err != nil {
+					return nil, fmt.Errorf("failed to set user's roles: %v", err)
+				}
 				return []*schema.ResourceData{d}, nil
 			},
 		},

--- a/okta/resource_okta_user.go
+++ b/okta/resource_okta_user.go
@@ -3,6 +3,7 @@ package okta
 import (
 	"context"
 	"fmt"
+	"reflect"
 
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
@@ -290,6 +291,56 @@ func resourceUser() *schema.Resource {
 				ValidateDiagFunc: stringLenBetween(4, 1000),
 				Description:      "User Password Recovery Answer",
 			},
+			"password_hash": {
+				Type:        schema.TypeSet,
+				MaxItems:    1,
+				Description: "Specifies a hashed password to import into Okta.",
+				Optional:    true,
+				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+					oldHash, newHash := d.GetChange("password_hash")
+					if oldHash != nil && newHash != nil && len(oldHash.(*schema.Set).List()) > 0 && len(newHash.(*schema.Set).List()) > 0 {
+						oh := oldHash.(*schema.Set).List()[0].(map[string]interface{})
+						nh := newHash.(*schema.Set).List()[0].(map[string]interface{})
+						return reflect.DeepEqual(oh, nh)
+					}
+					return new == "" || old == new
+				},
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"algorithm": {
+							Description:      "The algorithm used to generate the hash using the password",
+							Type:             schema.TypeString,
+							Required:         true,
+							ValidateDiagFunc: elemInSlice([]string{"BCRYPT", "SHA-512", "SHA-256", "SHA-1", "MD5"}),
+						},
+						"work_factor": {
+							Description:      "Governs the strength of the hash and the time required to compute it. Only required for BCRYPT algorithm",
+							Type:             schema.TypeInt,
+							Optional:         true,
+							ValidateDiagFunc: intBetween(1, 20),
+						},
+						"salt": {
+							Description: "Only required for salted hashes",
+							Type:        schema.TypeString,
+							Optional:    true,
+						},
+						"salt_order": {
+							Description:      "Specifies whether salt was pre- or postfixed to the password before hashing",
+							Type:             schema.TypeString,
+							Optional:         true,
+							ValidateDiagFunc: elemInSlice([]string{"PREFIX", "POSTFIX"}),
+						},
+						"value": {
+							Description: "For SHA-512, SHA-256, SHA-1, MD5, This is the actual base64-encoded hash of the password (and salt, if used). This is the " +
+								"Base64 encoded value of the SHA-512/SHA-256/SHA-1/MD5 digest that was computed by either pre-fixing or post-fixing the salt to the " +
+								"password, depending on the saltOrder. If a salt was not used in the source system, then this should just be the the Base64 encoded " +
+								"value of the password's SHA-512/SHA-256/SHA-1/MD5 digest. For BCRYPT, This is the actual radix64-encoded hashed password.",
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
 		},
 	}
 }
@@ -307,6 +358,7 @@ func resourceUserCreate(ctx context.Context, d *schema.ResourceData, m interface
 	uc := &okta.UserCredentials{
 		Password: &okta.PasswordCredential{
 			Value: d.Get("password").(string),
+			Hash:  buildPasswordCredentialHash(d.Get("password_hash")),
 		},
 	}
 	recoveryQuestion := d.Get("recovery_question").(string)
@@ -406,6 +458,7 @@ func resourceUserUpdate(ctx context.Context, d *schema.ResourceData, m interface
 	groupChange := d.HasChange("group_memberships")
 	userChange := hasProfileChange(d)
 	passwordChange := d.HasChange("password")
+	passwordHashChange := d.HasChange("password_hash")
 	recoveryQuestionChange := d.HasChange("recovery_question")
 	recoveryAnswerChange := d.HasChange("recovery_answer")
 
@@ -424,9 +477,18 @@ func resourceUserUpdate(ctx context.Context, d *schema.ResourceData, m interface
 		return diag.Errorf("Only the status of a DEPROVISIONED user can be updated, we detected other change")
 	}
 
-	if userChange {
+	if userChange || passwordHashChange {
 		profile := populateUserProfile(d)
-		userBody := okta.User{Profile: profile}
+		userBody := okta.User{
+			Profile: profile,
+		}
+		if passwordHashChange {
+			userBody.Credentials = &okta.UserCredentials{
+				Password: &okta.PasswordCredential{
+					Hash: buildPasswordCredentialHash(d.Get("password_hash")),
+				},
+			}
+		}
 		_, _, err := client.User.UpdateUser(ctx, d.Id(), userBody, nil)
 		if err != nil {
 			return diag.Errorf("failed to update user: %v", err)
@@ -493,6 +555,32 @@ func resourceUserUpdate(ctx context.Context, d *schema.ResourceData, m interface
 	return resourceUserRead(ctx, d, m)
 }
 
+func resourceUserDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	logger(m).Info("deleting user", "id", d.Id())
+	err := ensureUserDelete(ctx, d.Id(), d.Get("status").(string), getOktaClientFromMetadata(m))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	return nil
+}
+
+func buildPasswordCredentialHash(rawPasswordHash interface{}) *okta.PasswordCredentialHash {
+	if rawPasswordHash == nil || len(rawPasswordHash.(*schema.Set).List()) == 0 {
+		return nil
+	}
+	passwordHash := rawPasswordHash.(*schema.Set).List()
+	hash := passwordHash[0].(map[string]interface{})
+	wf, _ := hash["work_factor"].(int)
+	h := &okta.PasswordCredentialHash{
+		Algorithm:  hash["algorithm"].(string),
+		Value:      hash["value"].(string),
+		WorkFactor: int64(wf),
+	}
+	h.Salt, _ = hash["salt"].(string)
+	h.SaltOrder, _ = hash["salt_order"].(string)
+	return h
+}
+
 // Checks whether any profile keys have changed, this is necessary since the profile is not nested. Also, necessary
 // to give a sensible user readable error when they attempt to update a DEPROVISIONED user. Previously
 // this error always occurred when you set a user's status to DEPROVISIONED.
@@ -503,15 +591,6 @@ func hasProfileChange(d *schema.ResourceData) bool {
 		}
 	}
 	return false
-}
-
-func resourceUserDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
-	logger(m).Info("deleting user", "id", d.Id())
-	err := ensureUserDelete(ctx, d.Id(), d.Get("status").(string), getOktaClientFromMetadata(m))
-	if err != nil {
-		return diag.FromErr(err)
-	}
-	return nil
 }
 
 func ensureUserDelete(ctx context.Context, id, status string, client *okta.Client) error {

--- a/okta/resource_okta_user_admin_roles_test.go
+++ b/okta/resource_okta_user_admin_roles_test.go
@@ -1,6 +1,7 @@
 package okta
 
 import (
+	"fmt"
 	"testing"
 
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
@@ -14,6 +15,7 @@ func TestAccOktaUserAdminRoles_crud(t *testing.T) {
 	start := mgr.GetFixtures("basic.tf", ri, t)
 	update := mgr.GetFixtures("basic_update.tf", ri, t)
 	remove := mgr.GetFixtures("basic_removal.tf", ri, t)
+	resourceName := fmt.Sprintf("%s.test", userAdminRoles)
 
 	resource.Test(t, resource.TestCase{
 		PreCheck:          func() { testAccPreCheck(t) },
@@ -22,12 +24,21 @@ func TestAccOktaUserAdminRoles_crud(t *testing.T) {
 		Steps: []resource.TestStep{
 			{
 				Config: start,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "admin_roles.#", "2"),
+				),
 			},
 			{
 				Config: update,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "admin_roles.#", "3"),
+				),
 			},
 			{
 				Config: remove,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "admin_roles.#", "1"),
+				),
 			},
 		},
 	})

--- a/okta/resource_okta_user_test.go
+++ b/okta/resource_okta_user_test.go
@@ -185,7 +185,6 @@ func TestAccOktaUser_updateAllAttributes(t *testing.T) {
 					resource.TestCheckResourceAttr(resourceName, "last_name", "Smith"),
 					resource.TestCheckResourceAttr(resourceName, "login", email),
 					resource.TestCheckResourceAttr(resourceName, "email", email),
-					resource.TestCheckResourceAttr(resourceName, "admin_roles.#", "2"),
 				),
 			},
 			{
@@ -195,7 +194,6 @@ func TestAccOktaUser_updateAllAttributes(t *testing.T) {
 					resource.TestCheckResourceAttr(resourceName, "last_name", "Smith"),
 					resource.TestCheckResourceAttr(resourceName, "login", email),
 					resource.TestCheckResourceAttr(resourceName, "email", email),
-					resource.TestCheckResourceAttr(resourceName, "admin_roles.#", "1"),
 					resource.TestCheckResourceAttr(resourceName, "city", "New York"),
 					resource.TestCheckResourceAttr(resourceName, "cost_center", "10"),
 					resource.TestCheckResourceAttr(resourceName, "country_code", "US"),
@@ -367,22 +365,6 @@ func TestAccOktaUser_updateDeprovisioned(t *testing.T) {
 	})
 }
 
-func TestAccOktaUser_validRole(t *testing.T) {
-	rName := acctest.RandStringFromCharSet(10, acctest.CharSetAlphaNum)
-
-	resource.Test(t, resource.TestCase{
-		PreCheck:          func() { testAccPreCheck(t) },
-		ProviderFactories: testAccProvidersFactories,
-		CheckDestroy:      testAccCheckUserDestroy,
-		Steps: []resource.TestStep{
-			{
-				Config:      testOktaUserConfigValidRole(rName),
-				ExpectError: regexp.MustCompile("expected value to be one of 'SUPER_ADMIN', 'ORG_ADMIN', 'API_ACCESS_MANAGEMENT_ADMIN', 'APP_ADMIN', 'USER_ADMIN', 'MOBILE_ADMIN', 'READ_ONLY_ADMIN', 'HELP_DESK_ADMIN', 'REPORT_ADMIN', 'GROUP_MEMBERSHIP_ADMIN', got 'GROUP_ADMIN'"),
-			},
-		},
-	})
-}
-
 func TestAccOktaUser_loginUpdates(t *testing.T) {
 	ri := acctest.RandInt()
 	mgr := newFixtureManager(oktaUser)
@@ -427,7 +409,6 @@ func testAccCheckUserDestroy(s *terraform.State) error {
 func testOktaUserConfigInvalidCustomProfileAttribute(r string) string {
 	return fmt.Sprintf(`
 resource okta_user "test" {
-  admin_roles = ["APP_ADMIN", "USER_ADMIN"]
   first_name  = "TestAcc"
   last_name   = "%[1]s"
   login       = "testAcc-%[1]s@example.com"
@@ -445,24 +426,11 @@ resource okta_user "test" {
 func testOktaUserConfigUpdateDeprovisioned(r string) string {
 	return fmt.Sprintf(`
 resource okta_user "test" {
-  admin_roles = ["APP_ADMIN", "USER_ADMIN"]
   first_name  = "TestAcc"
   last_name   = "%[1]s"
   login       = "testAcc-%[1]s@example.com"
   status      = "DEPROVISIONED"
   email       = "hello@example.com"
-}
-`, r)
-}
-
-func testOktaUserConfigValidRole(r string) string {
-	return fmt.Sprintf(`
-resource okta_user "test" {
-  admin_roles = ["APP_ADMIN", "USER_ADMIN", "GROUP_ADMIN"]
-  first_name  = "TestAcc"
-  last_name   = "Smith"
-  login       = "testAcc-%[1]s@example.com"
-  email       = "testAcc-%[1]s@example.com"
 }
 `, r)
 }

--- a/okta/resource_okta_user_test.go
+++ b/okta/resource_okta_user_test.go
@@ -307,6 +307,45 @@ func TestAccOktaUser_statusDeprovisioned(t *testing.T) {
 	})
 }
 
+func TestAccOktaUserHashedPassword(t *testing.T) {
+	ri := acctest.RandInt()
+	mgr := newFixtureManager(oktaUser)
+	config := mgr.GetFixtures("password_hash.tf", ri, t)
+	configUpdated := mgr.GetFixtures("password_hash_updated.tf", ri, t)
+	resourceName := fmt.Sprintf("%s.test", oktaUser)
+	email := fmt.Sprintf("testAcc-%d@example.com", ri)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:          func() { testAccPreCheck(t) },
+		ProviderFactories: testAccProvidersFactories,
+		CheckDestroy:      testAccCheckUserDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: config,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "first_name", "TestAcc"),
+					resource.TestCheckResourceAttr(resourceName, "last_name", "Smith"),
+					resource.TestCheckResourceAttr(resourceName, "login", email),
+					resource.TestCheckResourceAttr(resourceName, "email", email),
+					resource.TestCheckResourceAttr(resourceName, "status", userStatusStaged),
+					resource.TestCheckResourceAttr(resourceName, "password_hash.0.algorithm", "SHA-512"),
+				),
+			},
+			{
+				Config: configUpdated,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "first_name", "TestAcc"),
+					resource.TestCheckResourceAttr(resourceName, "last_name", "Smith"),
+					resource.TestCheckResourceAttr(resourceName, "login", email),
+					resource.TestCheckResourceAttr(resourceName, "email", email),
+					resource.TestCheckResourceAttr(resourceName, "status", userStatusStaged),
+					resource.TestCheckResourceAttr(resourceName, "password_hash.0.algorithm", "BCRYPT"),
+				),
+			},
+		},
+	})
+}
+
 func TestAccOktaUser_updateDeprovisioned(t *testing.T) {
 	ri := acctest.RandInt()
 	mgr := newFixtureManager(oktaUser)

--- a/okta/user_schema.go
+++ b/okta/user_schema.go
@@ -173,10 +173,10 @@ func syncCustomUserSchema(d *schema.ResourceData, subschema *okta.UserSchemaAttr
 	if subschema.Items != nil {
 		_ = d.Set("array_type", subschema.Items.Type)
 		_ = d.Set("array_one_of", flattenOneOf(subschema.Items.OneOf))
-		_ = d.Set("array_enum", convertStringArrToInterface(subschema.Items.Enum))
+		_ = d.Set("array_enum", convertStringSliceToInterfaceSlice(subschema.Items.Enum))
 	}
 	if len(subschema.Enum) > 0 {
-		_ = d.Set("enum", convertStringArrToInterface(subschema.Enum))
+		_ = d.Set("enum", convertStringSliceToInterfaceSlice(subschema.Enum))
 	}
 	return setNonPrimitives(d, map[string]interface{}{
 		"one_of": flattenOneOf(subschema.OneOf),

--- a/okta/utils.go
+++ b/okta/utils.go
@@ -173,7 +173,10 @@ func createCustomNestedResourceImporter(fields []string, errMessage string) *sch
 	}
 }
 
-func convertStringArrToInterface(stringList []string) []interface{} {
+func convertStringSliceToInterfaceSlice(stringList []string) []interface{} {
+	if len(stringList) == 0 {
+		return nil
+	}
 	arr := make([]interface{}, len(stringList))
 	for i, str := range stringList {
 		arr[i] = str

--- a/sdk/idp_discovery_rule.go
+++ b/sdk/idp_discovery_rule.go
@@ -123,7 +123,10 @@ func (m *APISupplement) GetIdpDiscoveryRule(ctx context.Context, policyID, id st
 	if err != nil {
 		return nil, nil, err
 	}
-	rule := &IdpDiscoveryRule{}
-	resp, err := m.RequestExecutor.Do(ctx, req, rule)
-	return rule, resp, err
+	var rule *IdpDiscoveryRule
+	resp, err := m.RequestExecutor.Do(ctx, req, &rule)
+	if err != nil {
+		return nil, resp, err
+	}
+	return rule, resp, nil
 }

--- a/website/docs/r/app_user_schema.html.markdown
+++ b/website/docs/r/app_user_schema.html.markdown
@@ -4,14 +4,31 @@ page_title: 'Okta: okta_app_user_schema'
 sidebar_current: 'docs-okta-resource-app-user-schema'
 description: |-
   Creates an Application User Schema property.
-  **DEPRECATED** use `okta_app_user_schema_property` instead.
 ---
 
 # okta_app_user_schema
 
+**DEPRECATED** use `okta_app_user_schema_property` instead.
+
 Creates an Application User Schema property.
 
 This resource allows you to create and configure a custom user schema property and associate it with an application.
+
+## Migrating to a renamed resource
+
+To migrate from `okta_app_user_schema.{my_name}` to `okta_app_user_schema_property.{my_name}` please follow these steps:
+
+1. Import the ` okta_app_user_schema_property.{my_name}`, but keep intact `okta_app_user_schema.{my_name}` (the old one).
+```
+terraform import okta_app_user_schema_property.{my_name} appID/myPropertyName
+````
+2. Copy-paste all the fields from `okta_app_user_schema.{my_name}` to `okta_app_user_schema_property.{my_name}`
+3. Remove resource with deprecated name from the state
+```
+terraform state rm okta_app_user_schema.{my_name}
+```
+4. Remove `okta_app_user_schema.{my_name}` resource from the `.tf` file manually.
+5. Run `terraform apply` to apply the default fields.
 
 ## Example Usage
 

--- a/website/docs/r/user.html.markdown
+++ b/website/docs/r/user.html.markdown
@@ -131,6 +131,17 @@ The following arguments are supported:
 
 - `recovery_answer` - (Optional) User password recovery answer.
 
+- `password hash` - (Optional) Specifies a hashed password to import into Okta. When updating a user with a hashed password the user must be in the `STAGED` status.  
+  - `algorithm"` - (Required) The algorithm used to generate the hash using the password (and salt, when applicable). Must be set to BCRYPT, SHA-512, SHA-256, SHA-1 or MD5.
+  - `salt` - (Optional) Only required for salted hashes. For BCRYPT, this specifies the radix64-encoded salt used to generate 
+  the hash, which must be 22 characters long. For other salted hashes, this specifies the base64-encoded salt used to generate the hash.
+  - `work_factor` - (Optional) Governs the strength of the hash and the time required to compute it. Only required for BCRYPT algorithm. Minimum value is 1, and maximum is 20.
+  - `salt_order` - (Optional) Specifies whether salt was pre- or postfixed to the password before hashing. Only required for salted algorithms.
+  - `value` - (OptionalFor SHA-512, SHA-256, SHA-1, MD5, This is the actual base64-encoded hash of the password (and salt, if used). 
+  This is the Base64 encoded value of the SHA-512/SHA-256/SHA-1/MD5 digest that was computed by either pre-fixing or post-fixing 
+  the salt to the password, depending on the saltOrder. If a salt was not used in the source system, then this should just be 
+  the Base64 encoded value of the password's SHA-512/SHA-256/SHA-1/MD5 digest. For BCRYPT, This is the actual radix64-encoded hashed password.
+
 ## Attributes Reference
 
 - `id` - (Optional) ID of the User schema property.

--- a/website/docs/r/user_schema.html.markdown
+++ b/website/docs/r/user_schema.html.markdown
@@ -4,14 +4,31 @@ page_title: 'Okta: okta_user_schema'
 sidebar_current: 'docs-okta-resource-user-schema'
 description: |-
   Creates a User Schema property.
-  **DEPRECATED** use `okta_user_schema_property` instead.
 ---
 
 # okta_user_schema
 
+**DEPRECATED** use `okta_user_schema_property` instead.
+
 Creates a User Schema property.
 
 This resource allows you to create and configure a custom user schema property.
+
+## Migrating to a renamed resource
+
+To migrate from `okta_user_schema.{my_name}` to `okta_user_schema_property.{my_name}` please follow these steps:
+
+1. Import the ` okta_user_schema_property.{my_name}`, but keep intact `okta_user_schema.{my_name}` (the old one).
+```
+terraform import okta_user_schema_property.{my_name} appID/myPropertyName
+````
+2. Copy-paste all the fields from `okta_user_schema.{my_name}` to `okta_user_schema_property.{my_name}`
+3. Remove resource with deprecated name from the state
+```
+terraform state rm okta_user_schema.{my_name}
+```
+4. Remove `okta_user_schema.{my_name}` resource from the `.tf` file manually.
+5. Run `terraform apply` to apply the default fields.
 
 ## Example Usage
 

